ikarus/finiteelements/mechanics/kirchhoffloveshell.hh

```diff
@@ -180,30 +203,179 @@ namespace Ikarus {
      * @param local The local coordinates at which results are to be calculated.
      * @param result The result container to store the calculated values.
      */
-    void calculateAt([[maybe_unused]] const ResultRequirementsType& req,
-                     [[maybe_unused]] const Dune::FieldVector<double, Traits::mydim>& local,
-                     [[maybe_unused]] ResultTypeMap<double>& result) const {
+    void calculateAt([[maybe_unused]] const ResultRequirementsType &req,
+                     [[maybe_unused]] const Dune::FieldVector<double, Traits::mydim> &local,
+                     [[maybe_unused]] ResultTypeMap<double> &result) const {
       DUNE_THROW(Dune::NotImplemented, "No results are implemented");
     }
 
     Dune::CachedLocalBasis<
         std::remove_cvref_t<decltype(std::declval<LocalView>().tree().child(0).finiteElement().localBasis())>>
         localBasis;
-    std::function<Eigen::Vector<double, Traits::worlddim>(const Eigen::Vector<double, Traits::worlddim>&,
-                                                          const double&)>
+    std::function<Eigen::Vector<double, Traits::worlddim>(const Dune::FieldVector<double, Traits::worlddim> &,
+                                                          const double &)>
         volumeLoad;
-    std::function<Eigen::Vector<double, Traits::worlddim>(const Eigen::Vector<double, Traits::worlddim>&,
-                                                          const double&)>
+    std::function<Eigen::Vector<double, Traits::worlddim>(const Dune::FieldVector<double, Traits::worlddim> &,
+                                                          const double &)>
         neumannBoundaryLoad;
-    const BoundaryPatch<GridView>* neumannBoundary;
+    const BoundaryPatch<GridView> *neumannBoundary;
     mutable Dune::BlockVector<Dune::RealTuple<double, Traits::dimension>> dispAtNodes;
+    DefaultMembraneStrain membraneStrain;
     double emod_;
     double nu_;
     double thickness_;
     size_t numberOfNodes{0};
     int order{};
 
   protected:
+    auto computeMaterialAndStrains(const Dune::FieldVector<double, 2> &gpPos, int gpIndex, const Geometry &geo,
+                                   const auto &uFunction) const {
+      using ScalarType = typename std::remove_cvref_t<decltype(uFunction)>::ctype;
+      using namespace Dune;
+      using namespace Dune::DerivativeDirections;
+      const auto [X, Jd, Hd]              = geo.impl().zeroFirstAndSecondDerivativeOfPosition(gpPos);
+      const auto J                        = toEigen(Jd);
+      const auto H                        = toEigen(Hd);
+      const Eigen::Matrix<double, 2, 2> A = J * J.transpose();
+      Eigen::Matrix<double, 3, 3> G;
```

Comment:
This could be `auto G = Eigen::Matrix<double, 3, 3>::Zero();`

Reply:
Also as we have `ScalarType` everywhere as static type, we should use it instead of `double?`.  Further down in the function this is actually done

Reply:
Actually ScalarType can be an auto diff type, which is used to construct derivatives wrt to the current nodal position to get forces and stiffness matrix. Since G is corresponding to the reference state it not derived wrt. thus it cannot be `ScalarType`